# Step 1 - Data Extraction

In [72]:
import pandas as pd

In [5]:
dataframe = pd.read_csv("ocorrencia_2010_2020_v2.csv")
dataframe

codigo_ocorrencia;codigo_ocorrencia2;ocorrencia_classificacao;ocorrencia_cidade;ocorrencia_uf;ocorrencia_aerodromo;ocorrencia_dia;ocorrencia_hora;total_recomendacoes
0     40211;40211;INCIDENTE;RIO DE JANEIRO;RJ;****;0...                                                                                                                   
1     40349;40349;INCIDENTE;BELÉM;PA;SBBE;03/01/2010...                                                                                                                   
2     40351;40351;INCIDENTE;RIO DE JANEIRO;RJ;SBRJ;0...                                                                                                                   
3     39527;39527;ACIDENTE;LUCAS DO RIO VERDE;MT;***...                                                                                                                   
4     40324;40324;INCIDENTE;PELOTAS;RS;SBPK;05/01/20...                                                                                                                   
...                                                 ...                                                                                                                   
5748  79757;79757;INCIDENTE GRAVE;LAGOA DA CONFUSÃO;...                                                                                                                   
5749  79802;79802;INCIDENTE;RIO DE JANEIRO;RJ;SBGL;3...                                                                                                                   
5750  79756;79756;INCIDENTE GRAVE;VICENTINA;MS;****;...                                                                                                                   
5751  79844;79844;INCIDENTE;RIO DE JANEIRO;RJ;SBJR;3...                                                                                                                   
5752  79844;79844;INCIDENTE;RIO DE JANEIRO;RJ;SBJR;0...                                                                                                                   

[5753 rows x 1 columns]

In [6]:
# separating the file by ";"

df = pd.read_csv("ocorrencia_2010_2020_v2.csv", sep=";")
df

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 ****     03/01/2010   
1                  BELÉM            PA                 SBBE     03/01/2010   
2         RIO DE JANEIRO            RJ                 SBRJ     03/01/2010   
3     LUCAS DO RIO VERDE            MT                 ****     04/01/2010   
4                PELOTAS            RS                 SBPK     05/01/2010   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 ****     30/12/2020   
5749      RIO DE JANEIRO            RJ                 SBGL     30/12/2020   
5750           VICENTINA            MS                 ****     31/12/2020   
5751      RIO DE JANEIRO            RJ                 SBJR     31/12/2020   
5752      RIO DE JANEIRO            RJ                 SBJR     01/01/2021   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [7]:
# checking dataset data types

df.dtypes

codigo_ocorrencia            int64
codigo_ocorrencia2           int64
ocorrencia_classificacao    object
ocorrencia_cidade           object
ocorrencia_uf               object
ocorrencia_aerodromo        object
ocorrencia_dia              object
ocorrencia_hora             object
total_recomendacoes          int64
dtype: object

In [8]:
# classifying column 'oocorrencia_dia' as type datetime

df2 = pd.read_csv("ocorrencia_2010_2020_v2.csv", sep=";", parse_dates=['ocorrencia_dia'])
df2.dtypes

codigo_ocorrencia                    int64
codigo_ocorrencia2                   int64
ocorrencia_classificacao            object
ocorrencia_cidade                   object
ocorrencia_uf                       object
ocorrencia_aerodromo                object
ocorrencia_dia              datetime64[ns]
ocorrencia_hora                     object
total_recomendacoes                  int64
dtype: object

In [9]:
df2.ocorrencia_dia.dt.month

0        3
1        3
2        3
3        4
4        5
        ..
5748    12
5749    12
5750    12
5751    12
5752     1
Name: ocorrencia_dia, Length: 5753, dtype: int64

# Step 2 - Data Validation

In [10]:
df2.head(10)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   
5              39807               39807                INCIDENTE   
6              40215               40215                INCIDENTE   
7              39707               39707          INCIDENTE GRAVE   
8              39156               39156          INCIDENTE GRAVE   
9              39711               39711          INCIDENTE GRAVE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                 ****     2010-03-01   
1               BELÉM            PA                 SBBE     2010-03-01   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-03-01   
3  LUCAS DO RIO VERDE            MT                 ****     2010-04-01   
4             PELOTAS            RS                 SBPK     2010-05-01   
5            SALVADOR            BA                 ****     2010-06-01   
6               COARI            AM                 SBUY     2010-07-01   
7            CANUTAMA            AM                 ****     2010-09-01   
8            CASCAVEL            PR                 SBCA     2010-10-01   
9       PARÁ DE MINAS            MG                 ****     2010-10-01   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0  
5        17:53:00                    0  
6        18:40:00                    0  
7        12:30:00                    3  
8        23:15:00                    2  
9        20:00:00                    0

In [11]:
# validation of date and month order in the dataset

df3 = pd.read_csv("ocorrencia_2010_2020_v2.csv", sep=";", parse_dates=['ocorrencia_dia'], dayfirst=True)
df3.head(10)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   
5              39807               39807                INCIDENTE   
6              40215               40215                INCIDENTE   
7              39707               39707          INCIDENTE GRAVE   
8              39156               39156          INCIDENTE GRAVE   
9              39711               39711          INCIDENTE GRAVE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                 ****     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   
5            SALVADOR            BA                 ****     2010-01-06   
6               COARI            AM                 SBUY     2010-01-07   
7            CANUTAMA            AM                 ****     2010-01-09   
8            CASCAVEL            PR                 SBCA     2010-01-10   
9       PARÁ DE MINAS            MG                 ****     2010-01-10   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0  
5        17:53:00                    0  
6        18:40:00                    0  
7        12:30:00                    3  
8        23:15:00                    2  
9        20:00:00                    0

## The Pandera library allows you to perform data validation, returning an error if the data does not agree with what we need.

In [12]:
import pandera as pa

In [13]:
schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia":pa.Column(pa.DateTime)
        
    }
)
schema.validate(df3)

SchemaError: expected series 'codigo_ocorrencia' to have type 'datetime64[ns]', got 'int64'

The error above returns that there is some invalid data in the day_occurrence column of the Dataframe.

In [14]:
# next, we will validate only one "codigo_ocorrencia" column for type Int

schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia":pa.Column(pa.Int)
        
    }
)

In [15]:
schema.validate(df3)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 ****     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 ****     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                 ****     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [16]:
# after verifying the successful validation, we can validate all columns of our DataFrame
schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String),
        "ocorrencia_aerodromo": pa.Column(pa.String),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)

In [17]:
schema.validate(df3)

SchemaError: non-nullable series 'ocorrencia_hora' contains null values: {4100: nan}

In [18]:
# correcting "ocorrencia_hora" column error / allowing null values

schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String),
        "ocorrencia_aerodromo": pa.Column(pa.String),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String, nullable=True),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)

In [19]:
schema.validate(df3)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 ****     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 ****     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                 ****     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

## Validating 24-hour time format

In [20]:
# creating a regular expression to handle the "ocorrencia_hora" column

schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String),
        "ocorrencia_aerodromo": pa.Column(pa.String),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String, pa.Check.str_matches(r'^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:[0-5][0-9])?$'), nullable=True),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)

In [21]:
schema.validate(df3)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 ****     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 ****     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                 ****     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

## Validating Character Size in a Column

In [22]:
'''
Some columns may require a limited number of values,
for example, the "ocorrencia_uf" column where is the abbreviation of a State.
'''

schema = pa.DataFrameSchema(
    columns = {
        "codigo": pa.Column(pa.Int),
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String, pa.Check.str_length(2,2)),
        "ocorrencia_aerodromo": pa.Column(pa.String),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String, pa.Check.str_matches(r'^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:[0-5][0-9])?$'), nullable=True),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)
schema.validate(df3)


SchemaError: column 'codigo' not in dataframe
   codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                 ****     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0  

In [23]:
# handling previous error with Require parameter: 

schema = pa.DataFrameSchema(
    columns = {
        "codigo": pa.Column(pa.Int, required=False),
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String, pa.Check.str_length(2,2)),
        "ocorrencia_aerodromo": pa.Column(pa.String),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String, pa.Check.str_matches(r'^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:[0-5][0-9])?$'), nullable=True),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)
schema.validate(df3)


codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 ****     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 ****     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                 ****     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

# Step 3 - Data Cleaning



- After validation, there is still data that is not ready to be Transformed.

Data such as #, ?, ! * must be removed for our DataFrame as they are not logically useful.

After checking the cvs spreadsheet, we identified that the files to be cleaned are:
ocorrencia_uf: (**)

ocorrencia_aerodromo: (##! - #### - ** - ***)

ocorrencia_hora: (NULL)

In [24]:
df3.head()

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                 ****     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                 ****     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0

In [25]:
'''
Clearing all ocorrencia_aerodromo lines that contain (****)
pa.NA = manipulates not evaluable values in a cell
'''

df3.loc[df3.ocorrencia_aerodromo == '****', ['ocorrencia_aerodromo']] = pd.NA

df3.head()

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                 <NA>     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                 <NA>     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0

In [26]:
# making changes to the Dataset as a whole

df3.replace(['**','***','****','*****','###!','####','NULL'], pd.NA, inplace=True)
df3

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                 <NA>     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                 <NA>     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                 <NA>     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                 <NA>     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [27]:
# check how much data is NOT AVAILABLE in the DataFrame

df3.isnull().sum()

codigo_ocorrencia              0
codigo_ocorrencia2             0
ocorrencia_classificacao       0
ocorrencia_cidade              0
ocorrencia_uf                  1
ocorrencia_aerodromo        2349
ocorrencia_dia                 0
ocorrencia_hora                1
total_recomendacoes            0
dtype: int64

In [28]:
# entering value 0 in NA data at preview time

df3.fillna(0)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                    0     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                    0     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                    0     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                    0     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [29]:
# entering value 0 in ALL NA data definitely

df3.fillna(0, inplace=True)

df3.head()

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                    0     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                    0     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0

In [30]:
# Checking if NOT AVAILABLE data has changed

df3.isnull().sum()

codigo_ocorrencia           0
codigo_ocorrencia2          0
ocorrencia_classificacao    0
ocorrencia_cidade           0
ocorrencia_uf               0
ocorrencia_aerodromo        0
ocorrencia_dia              0
ocorrencia_hora             0
total_recomendacoes         0
dtype: int64

In [31]:
df3

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                    0     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                    0     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                    0     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                    0     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [32]:
# excluding duplicate values 

df3.drop_duplicates(inplace=True)

df3


codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                    0     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                    0     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                    0     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                    0     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

# Step 4 - Data Transformation

In [73]:
# removing missing values

missing_values = ['**','****','*****','###!','####','NULL']

df3 = pd.read_csv("ocorrencia_2010_2020_v2.csv", sep=';', parse_dates=['ocorrencia_dia'], dayfirst=True, na_values=missing_values) 

df3.head(10)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0              40211               40211                INCIDENTE   
1              40349               40349                INCIDENTE   
2              40351               40351                INCIDENTE   
3              39527               39527                 ACIDENTE   
4              40324               40324                INCIDENTE   
5              39807               39807                INCIDENTE   
6              40215               40215                INCIDENTE   
7              39707               39707          INCIDENTE GRAVE   
8              39156               39156          INCIDENTE GRAVE   
9              39711               39711          INCIDENTE GRAVE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                  NaN     2010-01-03   
1               BELÉM            PA                 SBBE     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3  LUCAS DO RIO VERDE            MT                  NaN     2010-01-04   
4             PELOTAS            RS                 SBPK     2010-01-05   
5            SALVADOR            BA                  NaN     2010-01-06   
6               COARI            AM                 SBUY     2010-01-07   
7            CANUTAMA            AM                  NaN     2010-01-09   
8            CASCAVEL            PR                 SBCA     2010-01-10   
9       PARÁ DE MINAS            MG                  NaN     2010-01-10   

  ocorrencia_hora  total_recomendacoes  
0        12:00:00                    0  
1        11:05:00                    0  
2        03:00:00                    0  
3        17:30:00                    0  
4        19:25:00                    0  
5        17:53:00                    0  
6        18:40:00                    0  
7        12:30:00                    3  
8        23:15:00                    2  
9        20:00:00                    0

In [34]:
# validating schema

schema = pa.DataFrameSchema(
    columns = {
        "codigo_ocorrencia": pa.Column(pa.Int),
        "codigo_ocorrencia2": pa.Column(pa.Int),
        "ocorrencia_classificacao": pa.Column(pa.String),
        "ocorrencia_cidade": pa.Column(pa.String),
        "ocorrencia_uf": pa.Column(pa.String, pa.Check.str_length(2,2), nullable=True),
        "ocorrencia_aerodromo": pa.Column(pa.String, nullable=True),
        "ocorrencia_dia": pa.Column(pa.DateTime),
        "ocorrencia_hora": pa.Column(pa.String, pa.Check.str_matches(r'^([0-1]?[0-9]|[2][0-3]):([0-5][0-9])(:[0-5][0-9])?$'), nullable=True),
        "total_recomendacoes": pa.Column(pa.Int),
    }
)

In [35]:
schema.validate(df3)

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
1                 40349               40349                INCIDENTE   
2                 40351               40351                INCIDENTE   
3                 39527               39527                 ACIDENTE   
4                 40324               40324                INCIDENTE   
...                 ...                 ...                      ...   
5748              79757               79757          INCIDENTE GRAVE   
5749              79802               79802                INCIDENTE   
5750              79756               79756          INCIDENTE GRAVE   
5751              79844               79844                INCIDENTE   
5752              79844               79844                INCIDENTE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                  NaN     2010-01-03   
1                  BELÉM            PA                 SBBE     2010-01-03   
2         RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
3     LUCAS DO RIO VERDE            MT                  NaN     2010-01-04   
4                PELOTAS            RS                 SBPK     2010-01-05   
...                  ...           ...                  ...            ...   
5748   LAGOA DA CONFUSÃO            TO                  NaN     2020-12-30   
5749      RIO DE JANEIRO            RJ                 SBGL     2020-12-30   
5750           VICENTINA            MS                  NaN     2020-12-31   
5751      RIO DE JANEIRO            RJ                 SBJR     2020-12-31   
5752      RIO DE JANEIRO            RJ                 SBJR     2021-01-01   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
1           11:05:00                    0  
2           03:00:00                    0  
3           17:30:00                    0  
4           19:25:00                    0  
...              ...                  ...  
5748        18:30:00                    0  
5749        00:54:00                    0  
5750        09:00:00                    0  
5751        13:24:00                    0  
5752        13:24:00                    0  

[5753 rows x 9 columns]

In [36]:
df3.dtypes

codigo_ocorrencia                    int64
codigo_ocorrencia2                   int64
ocorrencia_classificacao            object
ocorrencia_cidade                   object
ocorrencia_uf                       object
ocorrencia_aerodromo                object
ocorrencia_dia              datetime64[ns]
ocorrencia_hora                     object
total_recomendacoes                  int64
dtype: object

In [37]:
# searching for information from a row of our Dataframe

df3.loc[1]

codigo_ocorrencia                         40349
codigo_ocorrencia2                        40349
ocorrencia_classificacao              INCIDENTE
ocorrencia_cidade                         BELÉM
ocorrencia_uf                                PA
ocorrencia_aerodromo                       SBBE
ocorrencia_dia              2010-01-03 00:00:00
ocorrencia_hora                        11:05:00
total_recomendacoes                           0
Name: 1, dtype: object

In [38]:
# 

df3.iloc[1]

codigo_ocorrencia                         40349
codigo_ocorrencia2                        40349
ocorrencia_classificacao              INCIDENTE
ocorrencia_cidade                         BELÉM
ocorrencia_uf                                PA
ocorrencia_aerodromo                       SBBE
ocorrencia_dia              2010-01-03 00:00:00
ocorrencia_hora                        11:05:00
total_recomendacoes                           0
Name: 1, dtype: object

In [39]:
# searching information from an last index of our Dataframe

df3.iloc[-1]

codigo_ocorrencia                         79844
codigo_ocorrencia2                        79844
ocorrencia_classificacao              INCIDENTE
ocorrencia_cidade                RIO DE JANEIRO
ocorrencia_uf                                RJ
ocorrencia_aerodromo                       SBJR
ocorrencia_dia              2021-01-01 00:00:00
ocorrencia_hora                        13:24:00
total_recomendacoes                           0
Name: 5752, dtype: object

In [40]:
# searching information from an column of our Dataframe

df3.loc[:, 'ocorrencia_uf']

0       RJ
1       PA
2       RJ
3       MT
4       RS
        ..
5748    TO
5749    RJ
5750    MS
5751    RJ
5752    RJ
Name: ocorrencia_uf, Length: 5753, dtype: object

In [41]:
# searching null values

df3.isnull().sum()

codigo_ocorrencia              0
codigo_ocorrencia2             0
ocorrencia_classificacao       0
ocorrencia_cidade              0
ocorrencia_uf                  1
ocorrencia_aerodromo        2349
ocorrencia_dia                 0
ocorrencia_hora                1
total_recomendacoes            0
dtype: int64

#####  Creating Filters

In [42]:
# creating a null value filter on the ocorrencia_uf column

filtro = df3.ocorrencia_uf.isnull()
df3.loc[filtro]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
2227              49474               49474                 ACIDENTE   

         ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
2227  ÁGUAS INTERNACIONAIS           NaN                  NaN     2013-09-02   

     ocorrencia_hora  total_recomendacoes  
2227        02:54:00                    0

In [43]:
#creating a null value filter on the ocorrencia_aerodromo

filtro = df3.ocorrencia_aerodromo.isnull()
df3.loc[filtro]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                 40211               40211                INCIDENTE   
3                 39527               39527                 ACIDENTE   
5                 39807               39807                INCIDENTE   
7                 39707               39707          INCIDENTE GRAVE   
9                 39711               39711          INCIDENTE GRAVE   
...                 ...                 ...                      ...   
5741              79754               79754          INCIDENTE GRAVE   
5744              79753               79753          INCIDENTE GRAVE   
5745              79755               79755                 ACIDENTE   
5748              79757               79757          INCIDENTE GRAVE   
5750              79756               79756          INCIDENTE GRAVE   

       ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0         RIO DE JANEIRO            RJ                  NaN     2010-01-03   
3     LUCAS DO RIO VERDE            MT                  NaN     2010-01-04   
5               SALVADOR            BA                  NaN     2010-01-06   
7               CANUTAMA            AM                  NaN     2010-01-09   
9          PARÁ DE MINAS            MG                  NaN     2010-01-10   
...                  ...           ...                  ...            ...   
5741        NOVA MARINGÁ            MT                  NaN     2020-12-28   
5744            GOIATUBA            GO                  NaN     2020-12-29   
5745           MATO RICO            PR                  NaN     2020-12-29   
5748   LAGOA DA CONFUSÃO            TO                  NaN     2020-12-30   
5750           VICENTINA            MS                  NaN     2020-12-31   

     ocorrencia_hora  total_recomendacoes  
0           12:00:00                    0  
3           17:30:00                    0  
5           17:53:00                    0  
7           12:30:00                    3  
9           20:00:00                    0  
...              ...                  ...  
5741        13:00:00                    0  
5744        12:00:00                    0  
5745        10:30:00                    0  
5748        18:30:00                    0  
5750        09:00:00                    0  

[2349 rows x 9 columns]

In [44]:
# creating a null value filter on the ocorrencia_hora

filtro = df3.ocorrencia_hora.isnull()
df3.loc[filtro]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
4100              78823               78823                INCIDENTE   

     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
4100           CORUMBÁ            MS                  NaN     2017-07-09   

     ocorrencia_hora  total_recomendacoes  
4100             NaN                    0

In [45]:
# calculating Null Values within Dataframe

df3.count()

codigo_ocorrencia           5753
codigo_ocorrencia2          5753
ocorrencia_classificacao    5753
ocorrencia_cidade           5753
ocorrencia_uf               5752
ocorrencia_aerodromo        3404
ocorrencia_dia              5753
ocorrencia_hora             5752
total_recomendacoes         5753
dtype: int64

In the ocorrencia_uf column we have the value 5752 and in the ocorrencia_aerodromo column 3404.

This happens because the count does not count null values in the Data Frame

In [74]:
# occurrences with more than 10 recommendations

filter_recommendations = df3.total_recomendacoes >10
df.loc[filter_recommendations]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
63                39992               39992                 ACIDENTE   
326               42250               42250                 ACIDENTE   
530               43471               43471                 ACIDENTE   
727               44377               44377                 ACIDENTE   
872               44796               44796                 ACIDENTE   
884               44888               44888          INCIDENTE GRAVE   
1062              45554               45554          INCIDENTE GRAVE   
1972              47938               47938                INCIDENTE   
2795              52265               52265                 ACIDENTE   
3930              66432               66432          INCIDENTE GRAVE   

         ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
63               SÃO PAULO            SP                 SBMT     17/02/2010   
326         RIO DE JANEIRO            RJ                 SBRJ     12/08/2010   
530              SÃO PAULO            SP                 ****     14/12/2010   
727               ITÁPOLIS            SP                 SDIO     09/04/2011   
872                 RECIFE            PE                 SBRF     13/07/2011   
884   SÃO JOSÉ DOS PINHAIS            PR                 SBCT     15/07/2011   
1062            UBERLÂNDIA            MG                 SBUL     15/11/2011   
1972              BRASÍLIA            DF                 SBBR     13/04/2013   
2795                SANTOS            SP                 ****     13/08/2014   
3930               VITÓRIA            ES                 ****     21/02/2017   

     ocorrencia_hora  total_recomendacoes  
63          20:07:00                   11  
326         12:26:00                   19  
530         20:15:00                   11  
727         14:50:00                   11  
872         09:54:00                   23  
884         17:05:00                   11  
1062        02:15:00                   20  
1972        18:00:00                   11  
2795        13:03:00                   13  
3930        11:47:00                   12

In [78]:
# occurrence cities with more than 10 recommendations

filter_recommendations_cities = df3.total_recomendacoes > 10
df3.loc [filter_recommendations_cities, 'ocorrencia_cidade']


63                 SÃO PAULO
326           RIO DE JANEIRO
530                SÃO PAULO
727                 ITÁPOLIS
872                   RECIFE
884     SÃO JOSÉ DOS PINHAIS
1062              UBERLÂNDIA
1972                BRASÍLIA
2795                  SANTOS
3930                 VITÓRIA
Name: ocorrencia_cidade, dtype: object

In [79]:
# checking list of cities and total of commendations

filter_recommendations_cidades = df3.total_recomendacoes > 10
df3.loc [filter_recommendations_cities, ['ocorrencia_cidade', 'total_recomendacoes']]

ocorrencia_cidade  total_recomendacoes
63               SÃO PAULO                   11
326         RIO DE JANEIRO                   19
530              SÃO PAULO                   11
727               ITÁPOLIS                   11
872                 RECIFE                   23
884   SÃO JOSÉ DOS PINHAIS                   11
1062            UBERLÂNDIA                   20
1972              BRASÍLIA                   11
2795                SANTOS                   13
3930               VITÓRIA                   12

In [55]:
# filtering Serious Incident Type Occurrences

filtro_incidentes_graves = df3.ocorrencia_classificacao =='INCIDENTE GRAVE'
df3.loc [filtro_incidentes_graves, ['ocorrencia_cidade', 'total_recomendacoes']]

ocorrencia_cidade  total_recomendacoes
7              CANUTAMA                    3
8              CASCAVEL                    2
9         PARÁ DE MINAS                    0
29             CURITIBA                    2
35               PALMAS                    0
...                 ...                  ...
5741       NOVA MARINGÁ                    0
5744           GOIATUBA                    0
5746      MANOEL URBANO                    0
5748  LAGOA DA CONFUSÃO                    0
5750          VICENTINA                    0

[737 rows x 2 columns]

In [80]:
# COMBINING TWO FILTERS - Serious Incidents and SP

filter1 = df3.ocorrencia_classificacao =='INCIDENTE GRAVE'
filter2 = df3.ocorrencia_uf == 'SP'
df3.loc [filter1 & filter2]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
136               40270               40270          INCIDENTE GRAVE   
171               40620               40620          INCIDENTE GRAVE   
219               41411               41411          INCIDENTE GRAVE   
297               42289               42289          INCIDENTE GRAVE   
314               42323               42323          INCIDENTE GRAVE   
...                 ...                 ...                      ...   
5591              79540               79540          INCIDENTE GRAVE   
5610              79566               79566          INCIDENTE GRAVE   
5704              79739               79739          INCIDENTE GRAVE   
5705              79705               79705          INCIDENTE GRAVE   
5715              79718               79718          INCIDENTE GRAVE   

          ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
136            PIRASSUNUNGA            SP                 SDPY     2010-04-15   
171               GUARULHOS            SP                 SBGR     2010-05-06   
219               SÃO PAULO            SP                 SBMT     2010-06-08   
297               SÃO PAULO            SP                 SBMT     2010-07-29   
314   SÃO JOSÉ DO RIO PRETO            SP                 SBSR     2010-08-05   
...                     ...           ...                  ...            ...   
5591              SÃO PAULO            SP                  NaN     2020-09-12   
5610          REGENTE FEIJÓ            SP                 SDYJ     2020-09-27   
5704              SÃO PAULO            SP                 SBSP     2020-12-03   
5705               SOROCABA            SP                 SDCO     2020-12-04   
5715              SÃO PAULO            SP                 SBMT     2020-12-11   

     ocorrencia_hora  total_recomendacoes  
136         20:15:00                    0  
171         15:50:00                    0  
219         19:30:00                    2  
297         20:06:00                    0  
314         12:01:00                    0  
...              ...                  ...  
5591        13:10:00                    0  
5610        19:33:00                    0  
5704        14:42:00                    0  
5705        15:30:00                    0  
5715        13:50:00                    0  

[137 rows x 9 columns]

In [81]:
# Occurrences whose (classification == SERIOUS INCIDENT OR INCIDENT) AND status == SP

filter3 = (df3.ocorrencia_classificacao == 'INCIDENTE GRAVE') | (df3.ocorrencia_classificacao == 'INCIDENTE')
filter4 = df3.ocorrencia_uf == 'SP'
df3.loc [filter3 & filter4]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
16                39809               39809                INCIDENTE   
18                39828               39828                INCIDENTE   
26                39847               39847                INCIDENTE   
27                39768               39768                INCIDENTE   
33                39848               39848                INCIDENTE   
...                 ...                 ...                      ...   
5704              79739               79739          INCIDENTE GRAVE   
5705              79705               79705          INCIDENTE GRAVE   
5715              79718               79718          INCIDENTE GRAVE   
5742              79800               79800                INCIDENTE   
5747              79804               79804                INCIDENTE   

     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
16           SÃO PAULO            SP                  NaN     2010-01-15   
18              SANTOS            SP                  NaN     2010-01-15   
26            SOROCABA            SP                  NaN     2010-01-20   
27            CAMPINAS            SP                  NaN     2010-01-21   
33           GUARULHOS            SP                 SBGR     2010-01-25   
...                ...           ...                  ...            ...   
5704         SÃO PAULO            SP                 SBSP     2020-12-03   
5705          SOROCABA            SP                 SDCO     2020-12-04   
5715         SÃO PAULO            SP                 SBMT     2020-12-11   
5742         SÃO PAULO            SP                 SBMT     2020-12-28   
5747          CAMPINAS            SP                 SBKP     2020-12-29   

     ocorrencia_hora  total_recomendacoes  
16          15:00:00                    0  
18          17:45:00                    0  
26          13:10:00                    0  
27          20:45:00                    0  
33          17:05:00                    0  
...              ...                  ...  
5704        14:42:00                    0  
5705        15:30:00                    0  
5715        13:50:00                    0  
5742        10:15:00                    0  
5747        19:00:00                    0  

[985 rows x 9 columns]

In [82]:
# using the "is in" function for the antior example

filter5 = df3.ocorrencia_classificacao.isin(['INCIDENTE GRAVE', 'INCIDENTE'])
filter6 = df3.ocorrencia_uf == 'SP'
df3.loc [filter5 & filter6]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
16                39809               39809                INCIDENTE   
18                39828               39828                INCIDENTE   
26                39847               39847                INCIDENTE   
27                39768               39768                INCIDENTE   
33                39848               39848                INCIDENTE   
...                 ...                 ...                      ...   
5704              79739               79739          INCIDENTE GRAVE   
5705              79705               79705          INCIDENTE GRAVE   
5715              79718               79718          INCIDENTE GRAVE   
5742              79800               79800                INCIDENTE   
5747              79804               79804                INCIDENTE   

     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
16           SÃO PAULO            SP                  NaN     2010-01-15   
18              SANTOS            SP                  NaN     2010-01-15   
26            SOROCABA            SP                  NaN     2010-01-20   
27            CAMPINAS            SP                  NaN     2010-01-21   
33           GUARULHOS            SP                 SBGR     2010-01-25   
...                ...           ...                  ...            ...   
5704         SÃO PAULO            SP                 SBSP     2020-12-03   
5705          SOROCABA            SP                 SDCO     2020-12-04   
5715         SÃO PAULO            SP                 SBMT     2020-12-11   
5742         SÃO PAULO            SP                 SBMT     2020-12-28   
5747          CAMPINAS            SP                 SBKP     2020-12-29   

     ocorrencia_hora  total_recomendacoes  
16          15:00:00                    0  
18          17:45:00                    0  
26          13:10:00                    0  
27          20:45:00                    0  
33          17:05:00                    0  
...              ...                  ...  
5704        14:42:00                    0  
5705        15:30:00                    0  
5715        13:50:00                    0  
5742        10:15:00                    0  
5747        19:00:00                    0  

[985 rows x 9 columns]

In [83]:
# filter cities starting with letter C

filter_c = df3.ocorrencia_cidade.str[0] == 'C'
df3.loc[filter_c]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
6                 40215               40215                INCIDENTE   
7                 39707               39707          INCIDENTE GRAVE   
8                 39156               39156          INCIDENTE GRAVE   
15                39315               39315                 ACIDENTE   
20                40310               40310                INCIDENTE   
...                 ...                 ...                      ...   
5661              79649               79649                INCIDENTE   
5663              79652               79652                INCIDENTE   
5680              79683               79683                INCIDENTE   
5696              79692               79692                 ACIDENTE   
5747              79804               79804                INCIDENTE   

     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
6                COARI            AM                 SBUY     2010-01-07   
7             CANUTAMA            AM                  NaN     2010-01-09   
8             CASCAVEL            PR                 SBCA     2010-01-10   
15            CANUTAMA            AM                  NaN     2010-01-15   
20     CRUZEIRO DO SUL            AC                 SBCZ     2010-01-18   
...                ...           ...                  ...            ...   
5661          CAMPINAS            SP                 SBKP     2020-11-04   
5663           CONFINS            MG                 SBCF     2020-11-06   
5680   CRUZEIRO DO SUL            AC                 SBCZ     2020-11-16   
5696         CATANDUVA            SP                 SDCD     2020-11-28   
5747          CAMPINAS            SP                 SBKP     2020-12-29   

     ocorrencia_hora  total_recomendacoes  
6           18:40:00                    0  
7           12:30:00                    3  
8           23:15:00                    2  
15          21:22:00                    0  
20          19:32:00                    0  
...              ...                  ...  
5661        22:39:00                    0  
5663        14:20:00                    0  
5680        22:00:00                    0  
5696        12:50:00                    0  
5747        19:00:00                    0  

[825 rows x 9 columns]

In [84]:
# Filter cities with letter endings

filter_ma = df3.ocorrencia_cidade.str[-2:] == 'MA'
df3.loc[filter_ma]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
7                 39707               39707          INCIDENTE GRAVE   
15                39315               39315                 ACIDENTE   
408               43062               43062                 ACIDENTE   
688               44198               44198                 ACIDENTE   
1035              51785               51785                INCIDENTE   
1377              45939               45939                 ACIDENTE   
1401              46180               46180                 ACIDENTE   
2099              49113               49113          INCIDENTE GRAVE   
2102              48799               48799                 ACIDENTE   
2655              51347               51347                INCIDENTE   
2855              52340               52340                 ACIDENTE   
2966              52832               52832                 ACIDENTE   
2997              52979               52979                 ACIDENTE   
3040              52984               52984                 ACIDENTE   
3092              53596               53596                INCIDENTE   
3157              53223               53223                 ACIDENTE   
3171              53265               53265                 ACIDENTE   
3189              53271               53271          INCIDENTE GRAVE   
3682              65414               65414          INCIDENTE GRAVE   
3725              65633               65633          INCIDENTE GRAVE   
3881              66265               66265                 ACIDENTE   
4229              77371               77371                 ACIDENTE   
4311              77612               77612                 ACIDENTE   
4574              78080               78080                 ACIDENTE   
4986              78867               78867                INCIDENTE   
5070              78923               78923          INCIDENTE GRAVE   
5080              78782               78782                 ACIDENTE   
5405              79282               79282                 ACIDENTE   
5440              79338               79338                INCIDENTE   
5474              79400               79400                INCIDENTE   

     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
7             CANUTAMA            AM                  NaN     2010-01-09   
15            CANUTAMA            AM                  NaN     2010-01-15   
408          NOVO GAMA            GO                  NaN     2010-09-25   
688           BURITAMA            SP                  NaN     2011-03-19   
1035          BURITAMA            SP                 SIBX     2011-10-25   
1377          UMUARAMA            PR                  NaN     2012-05-30   
1401         AURIFLAMA            SP                  NaN     2012-06-16   
2099         NOVA LIMA            MG                  NaN     2013-06-15   
2102          CANUTAMA            AM                  NaN     2013-06-18   
2655           DIADEMA            SP                  NaN     2014-05-04   
2855              FAMA            MG                  NaN     2014-09-20   
2966          BURITAMA            SP                  NaN     2014-12-14   
2997          IVINHEMA            MS                  NaN     2015-01-02   
3040         NOVA LIMA            MG                 SJLY     2015-02-01   
3092           ITAPEMA            SC                  NaN     2015-03-10   
3157         PINDORAMA            SP                  NaN     2015-04-22   
3171         PACARAIMA            RR                 SDMM     2015-05-05   
3189          UMUARAMA            PR                  NaN     2015-05-17   
3682         NOVA LIMA            MG                 SJLY     2016-07-08   
3725         NOVA LIMA            MG                  NaN     2016-08-31   
3881          BURITAMA            SP                  NaN     2017-01-13   
4229          CANUTAMA            AM                 SDKH     2017-11-02   
4311           ITAPEMA   

In [85]:
filter_maal = df3.ocorrencia_cidade.str.contains('MA | AL')
df3.loc[filter_maal]

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
110               40089               40089                 ACIDENTE   
130               40258               40258                INCIDENTE   
131               40327               40327                INCIDENTE   
133               40248               40248                INCIDENTE   
156               40572               40572                 ACIDENTE   
...                 ...                 ...                      ...   
5652              79645               79645                 ACIDENTE   
5659              79647               79647                INCIDENTE   
5679              79678               79678                 ACIDENTE   
5682              79684               79684                INCIDENTE   
5733              79759               79759                INCIDENTE   

                     ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo  \
110   VILA BELA DA SANTÍSSIMA TRINDADE            MT                  NaN   
130                       MONTE ALEGRE            PA                  NaN   
131                       PORTO ALEGRE            RS                  NaN   
133                       PORTO ALEGRE            RS                 SBPA   
156                       PORTO ALEGRE            RS                 SBPA   
...                                ...           ...                  ...   
5652                       ALTO ALEGRE            RR                  NaN   
5659                      PORTO ALEGRE            RS                 SBPA   
5679  VILA BELA DA SANTÍSSIMA TRINDADE            MT                  NaN   
5682                      PORTO ALEGRE            RS                 SBPA   
5733                      PORTO ALEGRE            RS                 SBPA   

     ocorrencia_dia ocorrencia_hora  total_recomendacoes  
110      2010-03-28        15:10:00                    0  
130      2010-04-12        18:10:00                    0  
131      2010-04-12        13:13:00                    0  
133      2010-04-13        18:47:00                    0  
156      2010-04-28        14:10:00                    0  
...             ...             ...                  ...  
5652     2020-10-28        12:25:00                    0  
5659     2020-11-01        18:20:00                    0  
5679     2020-11-16        14:15:00                    0  
5682     2020-11-17        20:18:00                    0  
5733     2020-12-21        23:14:00                    0  

[136 rows x 9 columns]

In [62]:
# turning two columns into a date and time

df3.ocorrencia_dia.astype(str) + ' ' + df3.ocorrencia_hora

0       2010-01-03 12:00:00
1       2010-01-03 11:05:00
2       2010-01-03 03:00:00
3       2010-01-04 17:30:00
4       2010-01-05 19:25:00
               ...         
5748    2020-12-30 18:30:00
5749    2020-12-30 00:54:00
5750    2020-12-31 09:00:00
5751    2020-12-31 13:24:00
5752    2021-01-01 13:24:00
Length: 5753, dtype: object

In [64]:
# grouping year and month

filtro1 = df3.ocorrencia_dia.dt.year == 2015
filtro2 = df3.ocorrencia_dia.dt.month == 12

df201503 = df3.loc[filtro1 & filtro2]

df201503

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
3428              53573               53573                INCIDENTE   
3429              60601               60601                INCIDENTE   
3430              53634               53634                INCIDENTE   
3431              53636               53636                INCIDENTE   
3432              53575               53575                INCIDENTE   
3433              60637               60637                INCIDENTE   
3434              53625               53625                 ACIDENTE   
3435              53626               53626                 ACIDENTE   
3436              53628               53628                 ACIDENTE   
3437              53629               53629                 ACIDENTE   
3438              53631               53631                INCIDENTE   
3439              60636               60636                INCIDENTE   
3440              53627               53627                INCIDENTE   
3441              60604               60604                INCIDENTE   
3442              60611               60611                INCIDENTE   
3443              60613               60613                 ACIDENTE   
3444              60629               60629                 ACIDENTE   
3445              53630               53630                INCIDENTE   
3446              60603               60603                INCIDENTE   
3447              60612               60612                 ACIDENTE   
3448              60783               60783                 ACIDENTE   
3449              53639               53639                INCIDENTE   
3450              60609               60609          INCIDENTE GRAVE   
3451              60616               60616                 ACIDENTE   
3452              60620               60620                 ACIDENTE   
3453              60645               60645                 ACIDENTE   
3454              60608               60608                INCIDENTE   
3455              60625               60625                 ACIDENTE   
3456              60605               60605                INCIDENTE   
3457              60610               60610                INCIDENTE   
3458              60602               60602                INCIDENTE   
3459              60615               60615                INCIDENTE   
3460              60640               60640                INCIDENTE   
3461              60641               60641                INCIDENTE   
3462              60632               60632          INCIDENTE GRAVE   
3463              60600               60600                INCIDENTE   
3464              60642               60642                INCIDENTE   
3465              60631               60631                 ACIDENTE   
3466              60649               60649                 ACIDENTE   

          ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
3428              GUARULHOS            SP                 SBGR     2015-12-01   
3429                 PALMAS            TO                 SBPJ     2015-12-01   
3430                 PALMAS            TO                 SBPJ     2015-12-02   
3431                JUNDIAÍ            SP                 SBJD     2015-12-02   
3432  CAMPOS DOS GOYTACAZES            RJ                 SBFS     2015-12-03   
3433         BELO HORIZONTE            MG                 SBBH     2015-12-03   
3434               TRINDADE            GO                  NaN     2015-12-06   
3435              AMERICANA            SP                 SDAI     2015-12-06   
3436                  AGUAÍ            SP                  NaN     2015-12-08   
3437                  JALES            SP                 SDJL     2015-12-08   
3438               CAMPINAS            SP                 SBKP     2015-12-08   
3439          CAXIAS DO SUL            RS                 SBCX     2015-12-08   
3440               TERESINA            PI                 SBTE     2015-

In [65]:
# grouping occurrence code

df201503.groupby(['ocorrencia_classificacao']).codigo_ocorrencia.count()

ocorrencia_classificacao
ACIDENTE           14
INCIDENTE          23
INCIDENTE GRAVE     2
Name: codigo_ocorrencia, dtype: int64

In [87]:
# grouping data from the Southeast region year 2010

filter_year = df3.ocorrencia_dia.dt.year == 2010
filter_region = df3.ocorrencia_uf.isin(['SP','MG','ES','RJ'])
dfSoutheast2010 = df3.loc[filter_year & filter_region]

dfSoutheast2010

codigo_ocorrencia  codigo_ocorrencia2 ocorrencia_classificacao  \
0                40211               40211                INCIDENTE   
2                40351               40351                INCIDENTE   
9                39711               39711          INCIDENTE GRAVE   
11               40069               40069                 ACIDENTE   
16               39809               39809                INCIDENTE   
..                 ...                 ...                      ...   
541              43722               43722                 ACIDENTE   
544              43548               43548                INCIDENTE   
548              43724               43724                 ACIDENTE   
550              43730               43730                INCIDENTE   
551              43975               43975                INCIDENTE   

    ocorrencia_cidade ocorrencia_uf ocorrencia_aerodromo ocorrencia_dia  \
0      RIO DE JANEIRO            RJ                  NaN     2010-01-03   
2      RIO DE JANEIRO            RJ                 SBRJ     2010-01-03   
9       PARÁ DE MINAS            MG                  NaN     2010-01-10   
11          SÃO PAULO            SP                 SBMT     2010-01-10   
16          SÃO PAULO            SP                  NaN     2010-01-15   
..                ...           ...                  ...            ...   
541            CAJOBI            SP                  NaN     2010-12-23   
544         SÃO PAULO            SP                 SBSP     2010-12-26   
548        COSMÓPOLIS            SP                  NaN     2010-12-29   
550           VITÓRIA            ES                  NaN     2010-12-30   
551   ARRAIAL DO CABO            RJ                  NaN     2010-12-30   

    ocorrencia_hora  total_recomendacoes  
0          12:00:00                    0  
2          03:00:00                    0  
9          20:00:00                    0  
11         14:50:00                    8  
16         15:00:00                    0  
..              ...                  ...  
541        11:00:00                    0  
544        22:10:00                    0  
548        12:00:00                    0  
550        18:10:00                    0  
551        14:28:00                    0  

[228 rows x 9 columns]

In [90]:
# ranking accidents

dfSoutheast2010.groupby(['ocorrencia_classificacao']).size()

ocorrencia_classificacao
ACIDENTE            40
INCIDENTE          170
INCIDENTE GRAVE     18
dtype: int64

In [91]:
# rank by state

dfSoutheast2010.groupby(['ocorrencia_classificacao', 'ocorrencia_uf']).size()

ocorrencia_classificacao  ocorrencia_uf
ACIDENTE                  MG                7
                          RJ                9
                          SP               24
INCIDENTE                 ES                5
                          MG               39
                          RJ               52
                          SP               74
INCIDENTE GRAVE           MG                5
                          RJ                4
                          SP                9
dtype: int64

In [92]:
# sorting descending order

dfSoutheast2010.groupby(['ocorrencia_cidade']).size().sort_values(ascending=False)

ocorrencia_cidade
RIO DE JANEIRO           47
SÃO PAULO                33
GUARULHOS                18
BELO HORIZONTE           17
CAMPINAS                 13
MACAÉ                    12
CONFINS                   7
UBERLÂNDIA                5
JUIZ DE FORA              4
VITÓRIA                   4
MARÍLIA                   3
SÃO JOSÉ DOS CAMPOS       3
SÃO JOSÉ DO RIO PRETO     3
SOROCABA                  2
ITANHAÉM                  2
RIBEIRÃO PRETO            2
IPATINGA                  2
POÇOS DE CALDAS           2
DIVINÓPOLIS               2
BAURU                     2
JUNDIAÍ                   2
ARARAQUARA                2
CAMPOS DOS GOYTACAZES     2
AREALVA                   2
MONTES CLAROS             2
UBERABA                   2
POUSO ALEGRE              1
PIRACICABA                1
PIRASSUNUNGA              1
ARRAIAL DO CABO           1
RESENDE                   1
PENÁPOLIS                 1
SANTOS                    1
SERRA NEGRA               1
SÃO BERNARDO DO CAMPO     1
TI